relayer/network_utils.go

```diff
@@ -52,33 +61,43 @@ func InitializeConnectionsAndCheckStake(
 // Connect to the validators of the source blockchain. For each destination blockchain,
 // verify that we have connected to a threshold of stake.
 func connectToNonPrimaryNetworkPeers(
+	ctx context.Context,
 	logger logging.Logger,
 	network peers.AppRequestNetwork,
 	cfg *config.Config,
 	sourceBlockchain *config.SourceBlockchain,
 ) error {
 	subnetID := sourceBlockchain.GetSubnetID()
 	network.TrackSubnet(subnetID)
-	connectedValidators, err := network.GetConnectedCanonicalValidators(subnetID)
-	if err != nil {
-		logger.Error(
-			"Failed to connect to canonical validators",
-			zap.String("subnetID", subnetID.String()),
-			zap.Error(err),
-		)
-		return err
-	}
 	for _, destination := range sourceBlockchain.SupportedDestinations {
 		blockchainID := destination.GetBlockchainID()
-		if ok, warpConfig, err := checkForSufficientConnectedStake(logger, cfg, connectedValidators, blockchainID); !ok {
-			logger.Warn(
-				"Failed to connect to a threshold of stake",
-				zap.String("destinationBlockchainID", blockchainID.String()),
-				zap.Uint64("connectedWeight", connectedValidators.ConnectedWeight),
-				zap.Uint64("totalValidatorWeight", connectedValidators.ValidatorSet.TotalWeight),
-				zap.Any("WarpConfig", warpConfig),
-			)
-			return err
+		for {
+			connectedValidators, err := network.GetConnectedCanonicalValidators(subnetID)
+			if err != nil {
+				logger.Error(
+					"Failed to connect to canonical validators",
+					zap.String("subnetID", subnetID.String()),
+					zap.Error(err),
+				)
+				return err
+			}
+			if ok, warpConfig, err := checkForSufficientConnectedStake(logger, cfg, connectedValidators, blockchainID); ok && err == nil {
```

Comment:
I think we should still return `err` if `err != nil` here. It only happens if the warp config is invalid. I think we should only be looping if we return `err == nil && !ok`

Same thing applies for `connectToPrimaryNetworkPeers()`